modules/api/src/main/java/org/apache/fluo/api/observer/ObserversFactory.java

```diff
@@ -0,0 +1,119 @@
+ * the License.
+ */
+
+package org.apache.fluo.api.observer;
+
+import java.util.function.BiConsumer;
+
+import org.apache.fluo.api.client.FluoClient;
+import org.apache.fluo.api.config.FluoConfiguration;
+import org.apache.fluo.api.config.SimpleConfiguration;
+import org.apache.fluo.api.data.Bytes;
+import org.apache.fluo.api.data.Column;
+import org.apache.fluo.api.metrics.MetricsReporter;
+import org.apache.fluo.api.observer.Observer.NotificationType;
+
+/**
+ * Fluo Workers use this class to create {@link Observer}s to process notifications. Implementations
+ * of this class should emit zero or more {@link Observer}s.
+ *
+ * <p>
+ * When Fluo is initialized {@link #getObservedColumns(Context, BiConsumer)} is called. The columns
+ * it emits are stored in Zookeeper. Transactions will use the columns stored in Zookeeper to
+ * determine when to set notifications. When Workers call
+ * {@link #createObservers(ObserverConsumer, Context)}, the columns emitted must be the same as
+ * those emitted during initialization. If this is not the case, then the worker will fail to start.
+ *
+ * @see FluoConfiguration#setObserversFactory(String)
+ * @since 1.1.0
+ */
+public interface ObserversFactory {
+
+  /**
+   * @since 1.1.0
+   */
+  interface Context {
+    /**
+     * @return A configuration object with application configuration like that returned by
+     *         {@link FluoClient#getAppConfiguration()}
+     */
+    SimpleConfiguration getAppConfiguration();
+
+    /**
+     * @return A {@link MetricsReporter} to report application metrics from observers.
+     */
+    MetricsReporter getMetricsReporter();
+  }
+
+  /**
+   * Observers are emitted to the worker using this interface. This interface also allows
+   * {@link Observer}s to be related to the columns that trigger them.
+   *
+   * @since 1.1.0
+   */
+  interface ObserverConsumer {
+    void accept(Column observedColumn, NotificationType ntfyType, Observer observer);
+
+    /**
+     * This method was created to allow Observers written as lambda to be passed {@link String}
+     * instead of {@link Bytes}.
+     * 
+     * <pre>
+     * <code>
+     *   void createObservers(ObserverConsumer obsConsumer, Context ctx) {
+     *     obsConsumer.accepts(someColumn, WEAK, (tx,row,col) -> {
+     *      //row is of type String
+     *     };
+     *   }
+     * </code>
+     * </pre>
+     */
+    void accepts(Column observedColumn, NotificationType ntfyType, StringObserver observer);
+  }
+
+  /**
+   * This is method is called by Fluo Workers to create observers to process notifications.
+   *
+   * <p>
+   * Observers emitted may be called concurrently by multiple threads to process different
+   * notifications. Observers should be tolerant of this.
+   *
+   * @param obsConsumer Emit an applications observers to this consumer.
+   */
+  void createObservers(ObserverConsumer obsConsumer, Context ctx);
```

Comment:
I'm not sure what you mean about it not being able to return a stream, since it's essentially the same as returning a set. Perhaps you thought I meant a stream of unpaired observers, when really I just meant a stream of whatever ObserverConsumer is consuming.

In any case, I agree with you on the boilerplate reduction, and returning a stream would be just as bad, if not significantly worse, than returning a set or a map.

Perhaps instead, the context itself can become an instance of ObserverConsumer? This simplifies the API a bit, and matches what MapReduce does for things like counters.
(And, if you wished, you could make ObserversFactory a `Consumer<Context>`, but that'd force a confusing method of `void accept(Context)` so I wouldn't recommend it).